design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,359 @@
+harder w.r.t. the cases mentioned in the Background:
+
+1. Managed Resources defined as `base` and traversed by `patches` might not yet
+   have been deployed when the Composition is being validated.
+2. Composite Resources specified by `CompositeTypeRef` and traversed by
+   `patches` might not yet have been deployed when the Composition is being
+   validated.
+3. We do not have all the required schemas locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+So, I propose to implement a Composition Validating Webhook by validating each
+`resource`'s `patches` and final rendered `Composed Resource`, querying the API
+Server to get the needed schemas.
+
+I suggest we avoid validating `bases`, that would require to implement
+validation ignoring required fields, as those could be set at a later stage by
+patches. Assuming we implement full `patches` type coverage, validating `bases`
+per se can be avoided, as any error should either be caught validating the
+patches or can be attributed to the `base` itself. But this could be considered
+at a later stage.
+
+To validate `patches`,  we will need to implement ad-hoc validation functions,
+because the current patching logic is not taking into consideration schemas, as
+it’s dealing directly with unstructured data, therefore not reporting any
+schema related error. The code implemented for
+[upbound/upjet][upject-validation-code] or
+[crossplane-contrib/crossplane-lint][crossplane-lint-validation-code] could be
+good starting points for this.
+
+To render the Composed Resources, we will need to refactor the Composite
+resources reconciliation loop, extracting the rendering logic to remove any
+side effect it currently has. Which is similar to what has been done in the LSP
+[implementation][lsp-validation-code]. This step requires all `patches` to have
+been applied in order to have the final Managed Resource to validate.
+
+In order to address the possible absence of Managed Resources,
+CompositeResourceDefinitions or any other required external resource at
+validation time we should allow explicitly setting one of 2 modes, `strict` and
+`loose`(default) . These should be configurable through an annotation, e.g.
+`crossplane.io/composition-validation-mode`, on a Composition or maybe even
+globally via a flag, and would respectively imply that missing resources should
+result in rejection or be ignored.
+
+I suggest, at least at the moment, to ignore and fail open if `alpha` features
```

Comment:
How would the validator report the failure when ignored?

Reply:
What do you mean? I meant that in case of alpha features being used by a Composition, if such feature is known to change the final rendered resources, e.g. Functions, the validator should not validate anything and "fail open" in the sense that it should short-circuit and just accept the resource. At least in the first iteration.

Reply:
The question is independent of alpha features indeed, when we use the `loose`(default) mode, are we planning to let user know that something is wrong (i.e. validation failed). With `strict`, it is clear that we reject the resource with proper error message and user knows what is wrong there.

Reply:
We could emit warnings, such as "... CRD was not available" or "... used, so no validation performed".